### ISSUE TITLE
FTP hangs on connect

### ISSUE THREAD

**Issue by cammellos** [miner/clj-ftp#27]
Hi,
thanks for the library!

We are facing an issue with connecting to a ftp server.

The connection seems to be getting stuck in the `open` function, waiting for a server reply and never times out.

It quite difficult to replicate as it's intermittent, but my understanding is that the FtpClient actually connects, but hangs in receiving data from the server.

We are using the `with-ftp` method and setting a `data-timeout-ms`, but it's only set after `open` has run.

As a temporary fix, I have modified the library to set a data timeout before actually connecting to the server. We are now checking whether that fixes the issue, which might take some time as we don't have a way to replicate this condition locally.

It looks like it might be this issue:
http://stackoverflow.com/questions/2125350/commons-net-ftp-deadlock

I am happy to send a PR if we are able to verify that setting the data timeout before connecting solves the issue, but I might need some input on how you would like to change the `open` function signature.

At the moment we changed it to:

```
(defn open
  ([url] (open url "UTF-8"))
  ([url control-encoding
    {:keys [data-timeout-ms
            control-keep-alive-timeout-sec
            control-keep-alive-reply-timeout-ms]
     :or {data-timeout-ms -1
          control-keep-alive-timeout-sec 300
          control-keep-alive-reply-timeout-ms 1000}}]
```

But we are happy to change it.

I have attached the thread stacktrace.

In the meantime, I was wondering if you have any input, did you come across any similar problem?

[stack.txt](https://github.com/miner/clj-ftp/files/390253/stack.txt)

Again, thanks for the library and sorry for the lengthy message!
cheeers
andrea
[stack.txt](https://github.com/miner/clj-ftp/files/390299/stack.txt)

**Comment by miner** [comment]
I haven't seen this problem before.  It sounds like you want to call .setDataTimeout in open before .connect to see if that fixes the problem.  That's worth a try.

However, if you're stuck in actually opening the connection, you probably need to use .setConnectTimeout (inherited from SocketClient).

I'm happy to accept changes to fix this, especially if you can provide a good test case.

**Comment by cammellos** [comment]
Hi Steve,
thanks for the reply,

I can confirm that it's not the connection timing out.

I have tried setting setDataTimeout but the problem persisted,

I have now tried using `setDefaultTimeout` before connecting and so far so
good, finger crossed :)

I will do a bit of investigation and once I can confirm that fixes it I
will start working on a PR.

thanks!
andrea

**Comment by cammellos** [comment]
Hi,
we have been using our fix in production for the past month and we had no
issues with it, I will try to push a pull request this weekend.

On Thu, Sep 22, 2016 at 12:23 PM, Oskar Danielsson <notifications@github.com

**Comment by rutchkiwi** [comment]
@cammellos I'm also experiencing this issue, thanks for your investigation into it! Did your fix work in production, and how did you fix it? Any help is appreciated, if I can fix this I might try and make that PR as well!

**Comment by cammellos** [comment]
Hi Viktor,
yes we have been running our code in production for about 3/4 months now
and never seen the issue again, I can get the code  back and send a PR or
at least upload it to gist, I will look at it probably tomorrow, I'll keep
you updated

**Comment by rutchkiwi** [comment]
Thanks a lot Andrea! Either the PR or the gist would be massively appreciated! :)

**Comment by miner** [comment]
Fixed in release version 0.3.9.
